File: tests/run_tier3.lua
-- run_tier3.lua
-- AE2-ES Tier 3 Extended Soak Test Runner
-- Runs: test_soak.lua (1K+ micro-jobs, saturation, ghost items)
--        test_timeslicescheduler.lua (profiling, cooperative multitasking)
-- Output: performance report (printed as JSON to stdout)
-- Detects: memory leaks, yield gaps > 4s, job crashes

local Assert = require("tests.helpers.assertions")
local MockEnv = require("tests.helpers.mock_env")
MockEnv.setup()

-- =========================================================================
-- Helper: math.round (not in standard Lua)
-- =========================================================================
local function round(value, decimals)
  local mult = 10 ^ (decimals or 0)
  return math.floor(value * mult + 0.5) / mult
end

-- =========================================================================
-- Helper: JSON encoder (minimal, no external deps)
-- =========================================================================
local function json_encode(value)
  local t = type(value)
  if t == "nil" then
    return "null"
  elseif t == "boolean" then
    return value and "true" or "false"
  elseif t == "number" then
    return string.format("%.3f", value)
  elseif t == "string" then
    return string.format("%q", value)
  elseif t == "table" then
    local parts = {}
    -- Detect array vs object
    local isArray = true
    local maxIdx = 0
    for k in pairs(value) do
      if type(k) ~= "number" or k < 1 or math.floor(k) ~= k then
        isArray = false
        break
      end
      if k > maxIdx then maxIdx = k end
    end
    if isArray and maxIdx > 0 then
      for i = 1, maxIdx do
        parts[#parts + 1] = json_encode(value[i])
      end
      return "[" .. table.concat(parts, ",") .. "]"
    else
      local keys = {}
      for k in pairs(value) do
        keys[#keys + 1] = tostring(k)
      end
      table.sort(keys)
      for _, k in ipairs(keys) do
        parts[#parts + 1] = string.format("%q:%s", k, json_encode(value[k]))
      end
      return "{" .. table.concat(parts, ",") .. "}"
    end
  else
    return string.format("%q", tostring(value))
  end
end

-- =========================================================================
-- Helper: get current timestamp string (ISO-ish)
-- =========================================================================
local function getTimestamp()
  return os.date("!%Y-%m-%dT%H:%M:%SZ")
end

-- =========================================================================
-- Main execution
-- =========================================================================
print("AE2-ES Tier 3 Extended Soak Test Suite")
print("==================================================")
print("Lua version: " .. _VERSION)
print("")

-- Reset test state
Assert.reset()

-- Measure baseline memory
collectgarbage("collect")
local memBefore = collectgarbage("count")

-- Track job crashes
local jobCrashes = 0

-- Start timer
local startTime = os.clock()

-- -----------------------------------------------------------------
-- Phase 1: Soak Test (memory stress, saturation, ghost items)
-- -----------------------------------------------------------------
print("--- Phase 1: Soak Test (test_soak.lua) ---")
local ok, err = pcall(function()
  require("tests.test_soak")
end)
if not ok then
  print("ERROR: Soak test crashed: " .. tostring(err))
  jobCrashes = jobCrashes + 1
end

-- Force GC and measure after soak
collectgarbage("collect")
local memAfterSoak = collectgarbage("count")

-- -----------------------------------------------------------------
-- Phase 2: Time-Slice Scheduler Profiling
-- -----------------------------------------------------------------
print("")
print("--- Phase 2: Scheduler Profiling (test_timeslicescheduler.lua) ---")
local ok2, err2 = pcall(function()
  require("tests.test_timeslicescheduler")
end)
if not ok2 then
  print("ERROR: Scheduler test crashed: " .. tostring(err2))
  jobCrashes = jobCrashes + 1
end

-- Force final GC
collectgarbage("collect")
local memAfter = collectgarbage("count")

-- End timer
local elapsed = os.clock() - startTime

-- =========================================================================
-- Collect test results
-- =========================================================================
local results = Assert.getResults()
local totalTests = #results
local totalAssertions = 0
local totalFailures = 0

for _, test in ipairs(results) do
  totalAssertions = totalAssertions + (test.assertions or 0)
  totalFailures = totalFailures + (test.failures or 0)
end

-- =========================================================================
-- Build performance report
-- =========================================================================
local memDelta = memAfter - memBefore
local memDeltaSoak = memAfterSoak - memBefore
-- Check for memory leak (> 100% growth — soak tests already validate
-- per-group flatness at 15%; this macroscopic check is a safety net)
local memoryLeakDetected = (memBefore > 0) and (memDelta > memBefore * 1.0)
local yieldGapOver4s = (elapsed > 4.0)
local allPassed = (totalFailures == 0) and (not memoryLeakDetected) and (not yieldGapOver4s) and (jobCrashes == 0)

local report = {
  tier = "Tier 3",
  timestamp = getTimestamp(),
  elapsed_seconds = round(elapsed, 3),
  gc_memory_before_kb = round(memBefore, 1),
  gc_memory_after_kb = round(memAfter, 1),
  gc_memory_delta_kb = round(memDelta, 1),
  gc_memory_after_soak_kb = round(memAfterSoak, 1),
  tests_run = totalTests,
  assertions = totalAssertions,
  failures = totalFailures,
  memory_leak_detected = memoryLeakDetected,
  yield_gap_over_4s = yieldGapOver4s,
  job_crashes = jobCrashes,
  success = allPassed,
}

-- =========================================================================
-- Output
-- =========================================================================
-- Print full test summary via Assert
Assert.summary()

-- Print JSON report
print("")
print("============================================")
print("  TIER 3 PERFORMANCE REPORT (JSON)")
print("============================================")
print("")
print("TIER3_REPORT:" .. json_encode(report))

-- Write to file as well (for artifact upload)
local file, writeErr = io.open("tier3-performance-report.json", "w")
if file then
  file:write(json_encode(report))
  file:close()
  print("Report written to tier3-performance-report.json")
else
  print("WARNING: Could not write report file: " .. tostring(writeErr))
end

-- Exit
if allPassed then
  print("")
  print("TIER 3 PASSED")
  os.exit(0)
else
  print("")
  if totalFailures > 0 then
    print(string.format("FAIL: %d test failure(s)", totalFailures))
  end
  if memoryLeakDetected then
    print(string.format("FAIL: Memory leak detected (%.1f KB growth)", memDelta))
  end
  if yieldGapOver4s then
    print(string.format("FAIL: Yield gap %.1fs exceeds 4s threshold", elapsed))
  end
  if jobCrashes > 0 then
    print(string.format("FAIL: %d job crash(es)", jobCrashes))
  end
  os.exit(1)
end

File: tests/run_tier3.py
#!/usr/bin/env python3
"""
AE2-ES Tier 3 Extended Soak Test Runner (Python/lupa bridge)

Runs the Tier 3 test suite:
  - test_soak.lua (1K micro-jobs, saturation stress, ghost items)
  - test_timeslicescheduler.lua (profiling)
  
Produces a performance report JSON artifact.
Detects: memory leaks, yield gaps > 4s, job crashes.
"""

import json
import sys
import os

project_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
os.chdir(project_root)

try:
    from lupa import LuaRuntime
except ImportError:
    print("ERROR: lupa is required. Install with: pip install lupa")
    sys.exit(1)

lua = LuaRuntime(unpack_returned_tuples=True)

# Set up package paths so Lua require() works
lua.execute("""
    package.path = "./src/?.lua;./lib/?.lua;./?.lua;./tests/?.lua;./tests/?/init.lua;" .. package.path
""")

# Provide bit32 for LuaJIT
lua.execute("""
    if not bit32 then
        bit32 = {
            bxor = function(a, b)
                local result = 0; local bit = 1
                while a > 0 or b > 0 do
                    if (a % 2) ~= (b % 2) then result = result + bit end
                    a = math.floor(a/2); b = math.floor(b/2); bit = bit * 2
                end
                return result
            end,
            band = function(a, b)
                local result = 0; local bit = 1
                while a > 0 and b > 0 do
                    if a % 2 == 1 and b % 2 == 1 then result = result + bit end
                    a = math.floor(a/2); b = math.floor(b/2); bit = bit * 2
                end
                return result
            end,
        }
    end
""")

# Load assertions and mock environment
try:
    lua.execute('require("tests.helpers.assertions")')
    print("Loaded assertions module OK")
except Exception as e:
    print(f"ERROR loading assertions: {e}")
    sys.exit(1)

try:
    lua.execute('require("tests.helpers.mock_env")')
    lua.execute('require("tests.helpers.mock_env").setup()')
    print("Loaded mock_env OK")
except Exception as e:
    print(f"ERROR loading mock_env: {e}")
    sys.exit(1)


def lua_table_to_dict(table):
    """Convert a lupa Lua table to a plain Python dict."""
    if table is None:
        return None
    return {key: table[key] for key in table}


# Run the tier3 Lua runner
print("\nAE2-ES Tier 3 Extended Soak Test Suite")
print("=" * 50)
print()

# Stub os.exit so dofile returns and _G.TIER3_REPORT remains readable
lua.execute("""
    os.exit = function(code)
        _G.TIER3_EXIT_CODE = code or 0
    end
""")

success = True
job_crashes = 0

try:
    lua.execute('dofile("tests/run_tier3.lua")')
except Exception as e:
    print(f"Tier 3 runner error: {e}")
    success = False
    job_crashes = 1

# Read performance report from Lua (soak-phase memory gate)
lua_report = lua_table_to_dict(lua.globals().TIER3_REPORT)
if lua_report is None:
    print("ERROR: TIER3_REPORT not set by Lua runner", file=sys.stderr)
    sys.exit(1)

memory_leak = bool(lua_report.get("memory_leak_detected", False))
yield_gap = bool(lua_report.get("yield_gap_over_4s", False))
failures = int(lua_report.get("failures", 0))
job_crashes = max(job_crashes, int(lua_report.get("job_crashes", 0)))
mem_delta_soak = float(lua_report.get("gc_memory_delta_soak_kb", 0))

report = dict(lua_report)
report["success"] = (
    success
    and failures == 0
    and not memory_leak
    and not yield_gap
    and job_crashes == 0
)

report_path = "tier3-performance-report.json"
with open(report_path, "w") as f:
    json.dump(report, f, indent=2)

print(f"\nTier 3 Report: {json.dumps(report, indent=2)}")
print(f"Report written to: {report_path}")

# Check failure conditions
if not success or failures > 0:
    print("ERROR: Tier 3 tests failed!", file=sys.stderr)
    sys.exit(1)

if memory_leak:
    print(
        f"FAIL: Memory leak detected during soak! ({mem_delta_soak:.1f} KB soak delta)",
        file=sys.stderr,
    )
    sys.exit(1)

if yield_gap:
    elapsed = float(lua_report.get("elapsed_seconds", 0))
    print(f"FAIL: Yield gap {elapsed:.1f}s exceeds 4s threshold!", file=sys.stderr)
    sys.exit(1)

if job_crashes > 0:
    print(f"FAIL: {job_crashes} job crash(es) detected!", file=sys.stderr)
    sys.exit(1)

print("All Tier 3 thresholds passed.")
sys.exit(0)
